### PR TITLE
8302154: Hidden classes created by LambdaMetaFactory can't be unloaded

### PR DESCRIPTION
**Pull request by simonis**
Prior to [JDK-8239384](https://bugs.openjdk.org/browse/JDK-8239384)/[JDK-8238358](https://bugs.openjdk.org/browse/JDK-8238358) LambdaMetaFactory has created VM-anonymous classes which could easily be unloaded once they were not referenced any more. Starting with JDK 15 and the new "hidden class" based implementation, this is not the case any more, because the hidden classes will be strongly tied to their defining class loader. If this is the default application class loader, these hidden classes can never be unloaded which can easily lead to Metaspace exhaustion (see the [test case in the JBS issue](https://bugs.openjdk.org/secure/attachment/102601/LambdaClassLeak.java)). This is a regression compared to previous JDK versions which some of our applications have been affected from when migrating to JDK 17.

The reason why the newly created hidden classes are strongly linked to their defining class loader is not clear to me. JDK-8239384 mentions it as an "implementation detail":

> *4. the lambda proxy class has the strong relationship with the class loader (that will share the VM metaspace for its defining loader - implementation details)*

From my current understanding the strong link between a hidden class created by `LambdaMetaFactory` and its defining class loader is not strictly required. In order to prevent potential OOMs and fix the regression compared the JDK 14 and earlier I propose to create these hidden classes without the `STRONG` option.

I'll be happy to add the test case as JTreg test to this PR if you think that would be useful.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302154](https://bugs.openjdk.org/browse/JDK-8302154): Hidden classes created by LambdaMetaFactory can't be unloaded

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12493/head:pull/12493` \
`$ git checkout pull/12493`

Update a local copy of the PR: \
`$ git checkout pull/12493` \
`$ git pull https://git.openjdk.org/jdk.git pull/12493/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12493`

View PR using the GUI difftool: \
`$ git pr show -t 12493`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12493.diff">https://git.openjdk.org/jdk/pull/12493.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12493#issuecomment-1424371468)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java [hunk 1]
```diff
@@ -380,9 +380,9 @@ public Void run() {
             Lookup lookup;
             if (useImplMethodHandle) {
                 lookup = caller.defineHiddenClassWithClassData(classBytes, implementation, !disableEagerInitialization,
-                                                               NESTMATE, STRONG);
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:383]: Please remove the import for this constant as well. It seems to be unused now.

**Review by JornVernee — COMMENTED** [review]
Your analysis in the JBS issue seems reasonable, but I'll wait for Mandy to respond here.

I looked at the original PR and don't see a discussion of `STRONG` there, but I very vaguely remember it being talked about as fixing a latent issue (I don't remember the details though).

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java (hunk 1 above)
Comment (simonis) [reply src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:383]: Will do (and will updated the copyright year as well :)

**Comment by simonis** [comment]
> Your analysis in the JBS issue seems reasonable, but I'll wait for Mandy to respond here.
> 
> I looked at the original PR and don't see a discussion of `STRONG` there, but I very vaguely remember it being talked about as being needed to fix a latent issue (I don't remember the details though).

Hi Jorn,

thanks for your quick response. I'd definitely want to hear Mandy's opinion on this as well :)

It also looks like the debug builds are running into an assertion during archiving. Currently looking into that as well...

**Comment by mlchung** [comment]
> 4. the lambda proxy class has the strong relationship with the class loader (that will share the VM metaspace for its defining loader - implementation details)

A lambda proxy class is the implementation of the lambda/method reference in class `C` and it's logical part of class `C` and therefore it has a strong relationship with the class loader.

The VM implementation creates one `ClassLoaderData` in the metaspace space for each "weak" hidden class, i.e. it consumes more C heap memory.  If a hidden class has a strong relationship with its defining loader, it should keep it strong.

**Comment by iklam** [comment]
Even in JDK 11, a lambda proxy classes that's referenced by the cpCache (i.e., from a resolved invokedynamic instruction associated with a lamda expression) is always kept alive. See test below.

So if I understand correctly, this patch will not affect lamda expressions in Java source code. It affects only direct calls to `LambdaMetafactory.metafactory()`. Is this correct?

```
public class LambdaGC {
    public static void main(String[] args) throws Throwable  {
      System.out.println("Entering LambdaGC");
      doit(() -> {
          Thread.dumpStack();
      });
      for (int i = 0; i < 10; i++) {
            System.gc();
      }
      System.out.println("Finish LambdaGC");
    }
    static void doit(Runnable r) {
        r.run();
    }
}

$ java11 -cp . -XX:+UnlockDiagnosticVMOptions -XX:+ShowHiddenFrames -Xlog:class+load -Xlog:class+unload LambdaGC | grep LambdaGC
[0.022s][info][class,load] LambdaGC source: file:/jdk3/tmp/
Entering LambdaGC
[0.024s][info][class,load] LambdaGC$$Lambda$1/0x0000000840060840 source: LambdaGC
java.lang.Exception: Stack trace
	at java.base/java.lang.Thread.dumpStack(Thread.java:1387)
	at LambdaGC.lambda$main$0(LambdaGC.java:5)
	at LambdaGC$$Lambda$1/0x0000000840060840.run(<Unknown>:1000000)
	at LambdaGC.doit(LambdaGC.java:13)
	at LambdaGC.main(LambdaGC.java:4)
Finish LambdaGC
```

**Comment by mlchung** [comment]
> Even in JDK 11, a lambda proxy classes that's referenced by the cpCache (i.e., from a resolved invokedynamic instruction associated with a lamda expression) is always kept alive. 

Thanks @iklam.   That's exactly what I am trying to confirm.

**Comment by simonis** [comment]
@iklam, I think your understanding is correct. While the bootstrap methods for Java Lambdas do call `LambdaMetafactory.metafactory()`, they store the resulting call site (for Lambdas a `BoundMethodHandle`) in the appendix slot of the constant pool cache entry of the invokedynamic bytecode. The `BoundMethodHandle` contains a reference to an instance of the generated lambda form (i.e. in your example `LambdaGC$$Lambda$1`). This is enough in order to keep `LambdaGC$$Lambda$1` alive and prevent its unloading.

Until now, `LambdaGC$$Lambda$1` was also strongly linked to its defining class loader, but I don't think that's necessary to keep it alive (because it is referenced from the call site which is referenced from the constant pool cache).

Running your example with `-Xlog:indy+methodhandles=debug` confirms this:
```
set_method_handle bc=186 appendix=0x000000062b821838 method=0x00000008000c7698 (local signature)
{method}
 - this oop:          0x00000008000c7698
 - method holder:     'java/lang/invoke/Invokers$Holder'
 ...
appendix: java.lang.invoke.BoundMethodHandle$Species_L 
 {0x000000062b821838} - klass: 'java/lang/invoke/BoundMethodHandle$Species_L'
  - ---- fields (total size 5 words):
  ...
  - final 'argL0' 'Ljava/lang/Object;' @32  a 'LambdaGC$$Lambda$1+0x0000000801000400'{0x000000062b81e080} (0xc5703c10)
```

**Comment by mlchung** [comment]
> LambdaGC$$Lambda$1 was also strongly linked to its defining class loader, but I don't think that's necessary to keep it alive (because it is referenced from the call site which is referenced from the constant pool cache).

True but it was decided at that time for JEP 371 to make it strong so that the lambda proxy classes will share the same `ClassLoaderData` metaspace as other classes defined by the same loader.   This reduces the memory footprint consumption.

The question is:  what are the use cases for direct call to `LambdaMetafactory` that existing applications depend on?

**Comment by tstuefe** [comment]
@simonis This may be a stupid question, but how exactly does your patch help with Metaspace consumption? I mean, how is the deallocation path with the patch now?

We still have the CLD loaded by the BootClassLoader, right? So it is still living in the metaspace arena of the bootclassloader, even if the Lambda class itself is collected earlier. How is this memory freed or reused?

If there were a way to prematurely collect its metadata via Metaspace::deallocate, then it could be reused at least by the bootloader itself.

-----

>True but it was decided at that time for JEP 371 to make it strong so that the lambda proxy classes will share the same ClassLoaderData metaspace as other classes defined by the same loader. This reduces the memory footprint consumption.

Expanding on what @mlchung wrote, this has been a nice effect of JEP 371. Metaspace intra-chunk fragmentation went down since all these pesky little CLDs went away. It also saved some C-heap as a side effect for the CLDs itself.

Example, committed metaspace for 100k lambdas:
JDK11: 271MB
JDK15: (+JEP371): 221MB
JDK17: (+JEP371 +JEP387): 189MB

**Comment by simonis** [comment]
Here are some numbers. I basically measured the overhead in Metaspace and native memory (i.e. `malloc()`) for 10_000 Lambda classes created with `LambdaMetafactory.metafactory()`. For the Metaspace numbers I took the "used" column for both "Class" and "Non-Class" areas of all class loaders (i.e. "Total Usage") reported by `jcmd VM.metaspace show-loaders`. This is basically the same like the "Total" "BlockSz" reported by `jcmd VM.classloader_stats`.

For native memory overhead I took the "malloc" numbers from the "Class" section in the output from `VM.native_memory`. I did run the test program with `-Xint -XX:+UseSerialGC -Xms512m -Xmx512m -XX:MetaspaceSize=100M -Xlog:gc -XX:NativeMemoryTracking=detail LambdaClassLeak 10000` to make sure I have no garbage collections except the explicit one triggered by me. 

The results in the first two columns are the difference between the numbers before and after creating the 10_000 Lambda classes and the results in the last two columns are the numbers after doing a `System.gc()` and potentially unloading the created Lambda classes (because they are all not referenced from the program any more). 

| JDK | Metaspace (kb)  | malloc (kb)  | - after GC -> | Metaspace (kb)  | malloc (kb) | 
|:---:| ---------------:| ------------:| ------------- | ---------------:| -----------:|
| 11  |          16_328 |        6_561 |               |               0 |         232 |
| 15  |          13_624 |          240 |               |          13_624 |          92 |
| 17  |          13_394 |         3_38 |               |          13_394 |          93 |
| tip |          13_451 |          337 |               |          13_451 |          93 |
| fix |          13_437 |        5_957 |               |               0 |          10 |

As you can see, with JDK 11, where Lambda classes are implemented by VM Anonymous classes, the 10_000 classes take up ~16mb of Metaspace and 6mb of native memory. The native memory is required for the `ClassLoaderData` structures because each VM Anonymous class lives in its own `ClassLoaderData`. After a GC, all the Metaspace and most of the native memory gets freed up.

Starting with JDK 15,  Lambda classes are now backed by the newly introduced "Hidden Classes". The Metaspace consumption slightly drops from ~16 to ~13mb (because the classes are a little smaller) and the native consumption drops significantly from ~6mb to 240kb. That's because the newly generated hidden classes are all strongly coupled to their defining class loader and therefor live in its `ClassLoaderData` section. While this saves ~6mb of native memory, it also prevents unloading of these classes.

These numbers haven't changed up to the current JDK development version. With the fix proposed in this PR, we basically trade about ~6mb of native memory (i.e. ~600 bytes per Lambda class) for the ability to easily unload such Lambda classes once they are not referenced any more.

I still think doing this change is reasonable because:
1. `LambdaMetafactory.metafactory()` does not mention that the classes is creates will not be unloadable. And they actually were unloadable up to JDK 11 while they were backed by VM anonymous classes. This is unexpected for users.
2. I'd argue that Metaspace is more "*valuable*" than native memory because it is limited by `-XX:MetaspaceSize` whereas native memory is only limited by the amount of available memory on the system. Also, growing Metaspace will trigger garbage collections which might be unexpected.
3. I think only "real" users of Lambda functions should pay the price for them and not users who don't need them any more.
4. Finally, the PR fixes a regression compared to the JDK 11 behavior. 

In general it seems to me that creating a new  `ClassLoaderData` for each hidden (or VM anonymous) class is a lot of overhead and was probably only done as "a hack" to simplify the implementation of class unloading for such classes. This might have been appropriate in the "early days" when we had just a few VM anonymous classes. But as the usage of `LambdaMetafactory.metafactory()` becomes more common it may make sense to revise the implementation such that for example all hidden classes defined by a class loader can live in a single `ClassLoaderData` structure. I think this should be technically possible although it would complicate the unloading logic (and should therefore be postponed to a later change).

**Comment by forax** [comment]
Hi Volker, in the future, we may change the implementation of lambdas again, by example Valhalla parametric generics, Isolated methods (JDK-8158765), JDK-8087219 or JDK-8001537,  may allow to re-use the same class file/method for all lambdas implementing the same interface. I fear that forcing unloading may block future refactoring.

**Comment by JornVernee** [comment]
> (because they are all not referenced from the program any more).

So, it sounds like this is testing a case where `LambdaMetafactory.metafactory` is being called directly?

I'd like to point out that, while I buy that people are doing this (I see enough of this on StackOverflow), `LambdaMetafactory` is a runtime API specifically designed to support a language feature. Calling it directly is an unintended use case. We should cater towards the intended use case instead (the language feature) which benefits from the strong tie to the class loader.

This change also seems to hurt the common case in favor of the rare case. The native memory cost should be multiplied by the ratio of use case that see a regression (lambdas) to the use cases that see an improvement (direct calls to LambdaMetafactory). I think it is safe to say that that ration is quite large. i.e. we seem to be trading one regression for another, far worse one.

**Comment by tstuefe** [comment]
Hi Volker,

just a short feedback since I'm entering vacation mode.

> In general it seems to me that creating a new `ClassLoaderData` for each hidden (or VM anonymous) class is a lot of overhead and was probably only done as "a hack" to simplify the implementation of class unloading for such classes. This might have been appropriate in the "early days" when we had just a few VM anonymous classes. But as the usage of `LambdaMetafactory.metafactory()` becomes more common it may make sense to revise the implementation such that for example all hidden classes defined by a class loader can live in a single `ClassLoaderData` structure. I think this should be technically possible although it would complicate the unloading logic (and should therefore be postponed to a later change).

Interesting numbers. I underestimated the impact of native allocations.

A side note, the amount of waste avoided by JEP 371 depends highly on the actual size of the lambda (precisely, how close a lambda would be to the chunk size). That's pretty much random, but nice to know since it can shift the numbers from almost nil overhead to almost twofold overhead.

>I'd argue that Metaspace is more "valuable" than native memory because it is limited by -XX:MetaspaceSize, whereas native memory is only limited by the amount of available memory on the system. Also, growing Metaspace will trigger garbage collections which might be unexpected.

I don't follow that logic, for several reasons:
- Metaspace is more easily reclaimed, since it returns memory to the OS eagerly (since JEP 387). C-Heap, otoh, sticks to the process unless you (atm) manually trim. Depends on the libc used, but with glibc - arguable the main platform - this is the state. We are working on C-heap autotrim, but it is not completed yet. 
So while Metaspace is limited, it recovers from spikes more easily.
- The metaspace GC threshold is an (impreceise) mechanism to trigger GC based on native resource consumption. That this is not affected by C-heap memory is just a side effect of an imperfect implementation - ideally, all native resources associated with a class loader - be it metaspace or C-heap - together would weight into the decision of when to do a GC. In fact, your numbers may be taken as a hint that maybe CLDs should live in Metaspace instead.

---

As I stated before, we do have the ability to prematurely repurpose metaspace before the loader dies via Metaspace::deallocate(). It would be nice if there were a mechanism to trigger this at least from Java. E.g. by making Lambdas closable.

**Comment by mlchung** [comment]
> Finally, the PR fixes a regression compared to the JDK 11 behavior.

This is an intended change that improves the C heap memory usage.   In addition, the spec of `LambdaMetafactory` has no guarantee of the spinned classes be unloaded independent of it defining loader.   This is indeed a behavioral change if `LambdaMetafactory` API is called directly instead of via `invokedynamic`.   I can create a CSR to retrofit this behavioral change.

**Comment by liach** [comment]
This "called directly" usage reminds me of another behavioral change around the time "original" access for lookup was introduced, where a library can no longer create SAM interface instances out of static methods, as the provided lookup had private access (previously full privilege) but couldn't have original access in newer Java.

**Comment by simonis** [comment]
> > (because they are all not referenced from the program any more).
> 
> So, it sounds like this is testing a case where `LambdaMetafactory.metafactory` is being called directly?
> 
> I'd like to point out that, while I buy that people are doing this (I see enough of this on StackOverflow), `LambdaMetafactory` is a runtime API specifically designed to support a language feature. Calling it directly is an unintended use case.

Sorry, but that's a really weird argument. You can't introduce a new public API and then blame people for using it! If the API is really only intended to support the internal implementation of a language feature than make it private and don't export it.

**Comment by simonis** [comment]
> > Finally, the PR fixes a regression compared to the JDK 11 behavior.
> 
> This is an intended change that improves the C heap memory usage. In addition, the spec of `LambdaMetafactory` has no guarantee of the spinned classes be unloaded independent of it defining loader. This is indeed a behavioral change if `LambdaMetafactory` API is called directly instead of via `invokedynamic`. I can create a CSR to retrofit this behavioral change.

Yes, please. I think that would be helpful.

**Comment by dholmes-ora** [comment]
> Sorry, but that's a really weird argument. You can't introduce a new public API and then blame people for using it! If the API is really only intended to support the internal implementation of a language feature than make it private and don't export it.

The spec states:

> API Note:
>   These linkage methods are designed to support the evaluation of lambda expressions and method references in the Java Language. 

The API has to be public as the call gets injected into application code.

**Comment by AlanBateman** [comment]
> You can't introduce a new public API and then blame people for using it! If the API is really only intended to support the internal implementation of a language feature than make it private and don't export it.

There are a number of APIs that exist to support the Java Language. LambdaMetafactory makes it clear in the first paragraph of its API docs that it supports the lambda expression and method reference expression features of the Java Language. If a Java compiler compiles these language constructs to invokedynamic instructions then the bootstrap methods have to be standard APIs, otherwise the class files won't work on all VM implementations. The APIs adding to j.l.runtime are similar. Maybe some advance tooling/framework might use them directly but it should be rare.

**Comment by simonis** [comment]
I hear your arguments although I don't agree :)

Can we at least get consensus on that the current design to create a new ClassLoaderData for each non-strongly linked Hidden Class just in order to enable simple unloading isn't the greatest design and should eventually be replaced by a more sophisticated implementation which allows non-strongly linked Hidden Classes to share a single ClassLoaderData but still enable unloading for them once they aren't referenced any more?

**Comment by coleenp** [comment]
The reason that non-strongly linked classes have their own ClassLoaderData is because it implements the property that when the class loader is no longer loaded, metadata for it can be removed at once.  Even though Metaspace has been redesigned to be less fragmented, this is simpler than removing selected entities inside metaspace.  Replacing it with a more sophisticated design would not an be an improvement.
So I have to disagree with your statement above.

**Comment by iklam** [comment]
@simonis I want to ask a basic question -- what is the reason for your code to call `LambdaMetafactory.metafactory()` directly? It looks like you want to implement so sort of dynamic dispatch. Can equivalent functionality be implemented by the app itself?

If there's a real need for such a style of programming, and it requires some sort of built-in support in by the JDK, maybe we should have a proper API instead of piggy-backing on `LambdaMetafactory.metafactory()`.

I think if you give us more background, we can make this a more productive discussion than focusing on "did we make the right decision N versions ago" without defining what "right" means :-)

I would suggest re-booting this decision in the mailing lists rather than continuing in this PR.

**Comment by dholmes-ora** [comment]
I also have to disagree with the statement. The unit of unloading is the ClassLoader.

**Comment by simonis** [comment]
> The reason that non-strongly linked classes have their own ClassLoaderData is because it implements the property that when the class loader is no longer loaded, metadata for it can be removed at once. Even though Metaspace has been redesigned to be less fragmented, this is simpler than removing selected entities inside metaspace. Replacing it with a more sophisticated design would not an be an improvement. So I have to disagree with your statement above.

How could it be not an improvement to save ~600 bytes of native memory per class? Sorry, but I start to have real problems to follow this discussion. First you don't want to change strongly linked Hidden Classes to weakly linked ones **because** of the native memory overhead and then you say that a more sophisticated implementation which will eliminate this overhead would **not** be an improvement. I'm puzzled.

**Comment by simonis** [comment]
> I also have to disagree with the statement. The unit of unloading is the ClassLoader.

Hidden classes are not normal classes. They are not defined, created or loaded by a class loader. The only reason why hidden classes maintain a link to a *defining class loader* is because they need it to resolve types used by the hidden class's own fields and methods.
 
Some references from [JEP 371: Hidden Classes](https://openjdk.org/jeps/371):

> Dynamically generated classes may only be needed for a limited time, so retaining them for the lifetime of the statically generated class might unnecessarily increase memory footprint. Existing workarounds for this situation, such as per-class class loaders, are cumbersome and inefficient.

> A hidden class can be unloaded when it is no longer reachable, or it can share the lifetime of a class loader so that it is unloaded only when the class loader is garbage collected

> A hidden class is not created by a class loader and has only a loose connection to the class loader deemed to be its defining loader. We can turn these facts to our advantage by allowing a hidden class to be unloaded even if its notional defining loader cannot be reclaimed by the garbage collector.

> By default, Lookup::defineHiddenClass will create a hidden class that can be unloaded regardless of whether its notional defining loader is still alive. That is, when all instances of the hidden class are reclaimed and the hidden class is no longer reachable, it may be unloaded even though its notional defining loader is still reachable. This behavior is useful when a language runtime creates a hidden class to serve multiple classes defined by arbitrary class loaders: The runtime will see an improvement in footprint and performance relative to both `ClassLoader::defineClass()` and Unsafe::defineAnonymousClass()`

The only reason why hidden classes created by `LambdaMetaFactory` are strongly linked to a class loader (at least I haven't heard any other argument until now in this thread) is to *save native memory* and not because it is *logically required*! It's fine for me if you don't want to fix that. I can just not understand why you are all still insisting that creating a new ClassLoaderData object per hidden class is such a great decision and fixing that would not be beneficial.

Hidden classes were designed to be light-weight and easily unloadable (see JEP references above). In the case of `LambdaMetaFactory` we unnecessarily link them strongly to a class loader just because the current implementation is too expensive otherwise. On a side note, the JDK already creates non-strongly linked hidden classes as well, e.g. for `java.lang.invoke.MethodHandles$Lookup.unreflect()`.

**Comment by forax** [comment]
Volker, i think you are mixing lambda metafactory and hidden classes.

Hidden Classes have been envision has a way to generate adapters at runtime for languages that run on the JVM.
They replace VM anonymous classes that were both unsafe and always not bound to a classloader.

Lambda metafactory is the entry point to create lambda proxy for Java (the language).
The current implementation is using hidden classes to represent lambda proxy. In the past, it was using VM anonymous classes that why lambdas were not tie to a classloader, the fact that the implementation was using VM anonymous class was leaking.

To unload a Java class, its clasloader has to be unreachable. I see no reason this behavior should not be the same for Java lambdas classes. I believe it's what David is saying too.

Now, for https://github.com/aws/aws-sdk-java-v2/issues/3701, you can fix it by using a ClassValue to cache the class and/or using a hidden class. For me, this has nothing to do with Java lambdas, i.e. nothing to do with the lambda metafactory.

**Comment by JornVernee** [comment]
I'd like to suggest maybe using another option for turning MethodHandles into interface instances, such as using `MethodHandleProxies` or just lambda expressions which capture the particular `MethodHandle`, which is an option if the target type is statically known:

```java
MethodHandle mh = ...
TargetType instance = (Widget w) -> {
    try {
        return (SomeType) mh.invokeExact(w);
    } catch(Throwable t) {
        throw new RuntimeException(t); // or something more nuanced
    }
};
```

Most importantly, the above doesn't generate a new class every time, it generates just one. It can be slower in specific situations due to lack of inlining, but in the common use case it's not that bad (I've measure the extra hop through the method handle to be 4ns in the past in microbenchmarks). You could measure the impact for the particular app. (I guess the question is: do you _really_ want a new class to be generated for each MethodHandle?)

**Comment by kaspernielsen** [comment]
In all the places I've seen LambdaMetaFactory used, it is because of performance over reflection/non-static method handles. See, for example, https://www.optaplanner.org/blog/2018/01/09/JavaReflectionButMuchFaster.html. I believe Optaplanner is still using it.

There is also quite a number of posts on StackOverflow on people trying to use LambdaMetafactory:
https://stackoverflow.com/search?q=LambdaMetafactory

**Comment by forax** [comment]
@kaspernielsen, i believe that now that hidden class + class data are part of the public API,
you do not need to use lambda metafactory directly. But it requires Java 17 not 8 or 11.

**Comment by kaspernielsen** [comment]
@forax I think what attracted people was that you didn't have to add any dependencies (ASM). It is also simpler than having to deal with class files. Maybe once we get the Classfile API people will reconsider. Still, LambdaMetaFactory is just a couple of lines of code. So will probably be hard to resist for some people.

**Comment by mlchung** [comment]
As Remi, others and I explained, whether lambda metafactory uses strong hidden classes or VM anonymous classes (weak hidden classes) is implementation details.   The lambda proxy class is a logical part of the caller class as it's the implementation of the lambda, i.e. strongly-linked with the defining class loader.

**Pull request closed without merging**